pkg/process/util/api/weighted_queue.go

```diff
@@ -0,0 +1,208 @@
+// a maximum weight.  If adding an item to the queue would violate either the max weight or max size, earlier items
+// are purged from the queue until there is room for the newest item.
+//
+// Items added to the queue have a weight and type.  When purging existing items to make room for new, items of the same
+// type being added will be removed first before moving on to other types.
+type WeightedQueue struct {
+	// dataAvailable is a channel used for communication between callers invoking the Add method and callers
+	// blocked on Poll.  If a caller calls Poll and the queue is empty, the call will block waiting for a send
+	// on dataAvailable.  When Add is invoked, it will perform a non-blocking send on dataAvailable to notify the caller
+	// blocked on Poll
+	dataAvailable chan struct{}
+	queue         *list.List
+	mu            sync.Mutex
+
+	maxSize       int
+	maxWeight     int64
+	currentWeight int64
+}
+
+// NewWeightedQueue returns a new WeightedQueue with the given maximum size & weight
+func NewWeightedQueue(maxSize int, maxWeight int64) *WeightedQueue {
+	return &WeightedQueue{
+		dataAvailable: make(chan struct{}),
+		queue:         list.New(),
+		maxSize:       maxSize,
+		maxWeight:     maxWeight,
+		currentWeight: 0,
+	}
+}
+
+// Len returns the number of items in the queue
+func (q *WeightedQueue) Len() int {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.queue.Len()
+}
+
+// Weight returns the current weight of the queue
+func (q *WeightedQueue) Weight() int64 {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.currentWeight
+}
+
+// Poll retrieves the head of the queue or blocks until an item is available.  The provided exit channel can be closed
+// to interrupt the blocking operation.  Returns the head of the queue and true or nil, false if the poll was
+// interrupted by the closing of the exit channel
+func (q *WeightedQueue) Poll(exit chan struct{}) (WeightedItem, bool) {
+	for {
+		select {
+		case <-exit:
+			return nil, false
+		default:
+
+		}
+
+		q.mu.Lock()
+
+		// If the queue is empty, wait for a signal on the dataAvailable channel
+		if q.queue.Len() == 0 {
+			q.mu.Unlock()
+
+			select {
+			case <-q.dataAvailable:
+				continue
+			case <-exit:
+				return nil, false
+			}
+		}
+
+		e := q.queue.Front()
+		item := e.Value.(WeightedItem)
+		q.queue.Remove(e)
+		q.currentWeight -= item.Weight()
+
+		q.mu.Unlock()
+
+		return item, true
+	}
+}
+
+// Add adds the item to the queue.
+func (q *WeightedQueue) Add(item WeightedItem) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// If the item won't fit, don't even bother trying
+	if item.Weight() > q.maxWeight {
+		return
+	}
+
+	newWeight := q.currentWeight + item.Weight()
+
+	if newWeight > q.maxWeight {
+		weightToRemove := newWeight - q.maxWeight
+		weightRemoved := int64(0)
+
+		// Try to find an item of the same type that we can expire
+		for iter := q.iterator(); iter.hasNext(); iter.next() {
+			if v := iter.value(); v.Type() == item.Type() {
```

Comment:
Sorry if I missed this, but I thought you have separate queues for each type in the collector? If it's not the case, shouldn't each queue be storing only 1 type of element? Then you could simply remove from the earliest instead of having 2 pass approach.

Reply:
There is one queue per _endpoint_, not type . So we have one queue for pod checks and one queue for process/container/network data.